### PR TITLE
Make persistent_term:put/2 and persistent_term:erase/1 yield

### PR DESCRIPTION
**Pull request by kjellwinblad**
Prior to this commit, the functions put/2 and erase/1 in the
persistent_term module did not yield even if the persistent term table
was large (the persistent term table is copied every time put/2 or
erase/1 is called). Furthermore, a call to put/2 or erase/1 did only
consume a single reduction. This commit fixes these problems.

### CODE REVIEW

**Comment by bjorng** [comment]
I will have a look at the code tomorrow.

You should delete this part of the documentation:

https://github.com/erlang/otp/blob/3d712ffc9d08dad3a42ea988869eaef8cd99b53b/erts/doc/src/persistent_term.xml#L123-L135

File: erts/emulator/beam/erl_bif_persistent.c [hunk 1]
```diff
@@ -187,102 +232,167 @@ void erts_init_bif_persistent_term(void)
 			  am_persistent_term, am_info_trap, 1,
 			  &persistent_term_info_trap);
 }
+/* Macro used for trapping in persistent_term_put_2 and
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Good catch. Fixed in the commit below.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 2]
```diff
@@ -187,102 +232,167 @@ void erts_init_bif_persistent_term(void)
...
+    long max_iterations;
+#define PUT_TRAP_CODE                                                   \
+    BIF_TRAP2(bif_export[BIF_persistent_term_put_2], BIF_P, state_mref, BIF_ARG_2)
+#define TRAPPING_COPY_TABLE_PUT(TABLE_DEST, OLD_TABLE, NEW_SIZE, COPY_TYPE, LOC_NAME) \
+    TRAPPING_COPY_TABLE(TABLE_DEST, OLD_TABLE, NEW_SIZE, COPY_TYPE, LOC_NAME, PUT_TRAP_CODE)
 
+#ifdef DEBUG
+        (void)ITERATIONS_PER_RED;
+        iterations_until_trap = max_iterations =
+            (1103515245 * (ERTS_BIF_REDS_LEFT(BIF_P)) + (Uint)&ctx + 12345)  % 227;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Yes. Good suggestion. Fixed in the commit below.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 3]
```diff
@@ -187,102 +232,167 @@ void erts_init_bif_persistent_term(void)
...
+            TRAPPING_COPY_TABLE_PUT(ctx->hash_table,
+                                    ctx->hash_table,
+                                    ctx->hash_table->allocated,
+                                    ERTS_PERSISTENT_TERM_CPY_NO_REHASH,
+                                    PUT2_TRAP_LOCATION_REPLACE_VALUE);
         }
     }
 
+    {
+    Uint term_size;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Fixed in the commit below.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 4]
```diff
@@ -349,26 +459,79 @@ BIF_RETTYPE persistent_term_get_2(BIF_ALIST_2)
...
+
 BIF_RETTYPE persistent_term_erase_1(BIF_ALIST_1)
 {
-    Eterm key = BIF_ARG_1;
-    HashTable* old_table;
-    HashTable* new_table;
-    Uint entry_index;
-    Eterm old_term;
-
+    static const Uint ITERATIONS_PER_RED = 1;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: I have set the value to 32.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 5]
```diff
@@ -740,65 +915,102 @@ lookup(HashTable* hash_table, Eterm key)
...
+        ctx->iterations_done = 0;
     }
-    new_table->first_to_delete = 0;
-    new_table->num_to_delete = 0;
-    erts_atomic_init_nob(&new_table->refc, (erts_aint_t)1);
+    ctx->new_table->first_to_delete = 0;
+    ctx->new_table->num_to_delete = 0;
+    erts_atomic_init_nob(&ctx->new_table->refc, (erts_aint_t)1);
+    {
+    HashTable* new_table = ctx->new_table;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Fixed in the commit below.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 6]
```diff
@@ -349,26 +459,79 @@ BIF_RETTYPE persistent_term_get_2(BIF_ALIST_2)
...
+        }
+    } else {
+        /* Save state in magic bin in case trapping is necessary */
+        Eterm* hp;
+        Binary* state_bin = erts_create_magic_binary(sizeof(ErtsPersistentTermErase1Context),
+                                                     persistent_term_erase_1_ctx_bin_dtor);
+        hp = HAlloc(BIF_P, ERTS_MAGIC_REF_THING_SIZE);
+        state_mref = erts_mk_magic_ref(&hp, &MSO(BIF_P), state_bin);
+        ctx = ERTS_MAGIC_BIN_DATA(state_bin);
+        /* IMPORTANT: The following two fields are used to detect if
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Thanks for pointing this out. I have done the change you suggested on line 465.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 7]
```diff
@@ -349,26 +459,79 @@ BIF_RETTYPE persistent_term_get_2(BIF_ALIST_2)
     BIF_RET(result);
 }
 
+static int persistent_term_erase_1_ctx_bin_dtor(Binary *context_bin) {
+    ErtsPersistentTermErase1Context* ctx = ERTS_MAGIC_BIN_DATA(context_bin);
+    if (ctx->cpy_ctx.new_table != NULL) {
+        if (ctx->cpy_ctx.copy_type == ERTS_PERSISTENT_TERM_CPY_TEMP) {
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Good suggestion. I have used it in the commit below.

**Comment by kjellwinblad** [comment]
@bjorng: Thanks for the review. All the issues that you found should be fixed in commit 824fc64ac7717f0f7ccb4e94b291deba40a40e3d.

> You should delete this part of the documentation:

Done: https://github.com/erlang/otp/blob/824fc64ac7717f0f7ccb4e94b291deba40a40e3d/erts/doc/src/persistent_term.xml#L122

File: erts/emulator/beam/erl_bif_persistent.c [hunk 8]
```diff
@@ -740,65 +937,104 @@ lookup(HashTable* hash_table, Eterm key)
...
-        new_table->allocated = new_size;
-        new_table->num_entries = old_table->num_entries;
-        new_table->mask = new_size - 1;
-        for (i = 0; i < new_size; i++) {
-            new_table->term[i] = NIL;
+        ctx->new_table->allocated = ctx->new_size;
+        ctx->new_table->num_entries = ctx->old_table->num_entries;
+        ctx->new_table->mask = ctx->new_size - 1;
+    L_copy_table_place_2:
+        for (i = MAX(0, ctx->iterations_done);
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Good catch! Fixed in commit below.

File: erts/emulator/beam/erl_bif_persistent.c [hunk 9]
```diff
@@ -378,35 +561,49 @@ BIF_RETTYPE persistent_term_erase_1(BIF_ALIST_1)
...
+        new_size = ctx->tmp_table->allocated;
+        if (MUST_SHRINK(ctx->tmp_table)) {
             new_size /= 2;
         }
-        new_table = copy_table(tmp_table, new_size, 1);
-        erts_free(ERTS_ALC_T_TMP, tmp_table);
+        TRAPPING_COPY_TABLE_ERASE(ctx->new_table,
+                                  ctx->tmp_table,
+                                  new_size,
+                                  1,
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_bif_persistent.c]: Good catch! Fixed in commit below.

**Comment by kjellwinblad** [comment]
Merged

**Pull request closed without merging**